pkg/reconciler/addressableservice/addressableservice.go

```diff
@@ -18,115 +18,48 @@ package addressableservice
 type Reconciler struct {
-	// Client is used to write back status updates.
-	Client clientset.Interface
+	Core
 
 	// Listers index properties about resources
-	Lister        listers.AddressableServiceLister
 	ServiceLister corev1listers.ServiceLister
-
-	// The tracker builds an index of what resources are watching other
-	// resources so that we can immediately react to changes to changes in
-	// tracked resources.
-	Tracker tracker.Interface
-
-	// Recorder is an event recorder for recording Event resources to the
-	// Kubernetes API.
-	Recorder record.EventRecorder
 }
 
-// Check that our Reconciler implements controller.Reconciler
-var _ controller.Reconciler = (*Reconciler)(nil)
-
-// Reconcile implements controller.Reconciler
-func (r *Reconciler) Reconcile(ctx context.Context, key string) error {
-	logger := logging.FromContext(ctx)
-
-	// Convert the namespace/name string into a distinct namespace and name
-	namespace, name, err := cache.SplitMetaNamespaceKey(key)
-	if err != nil {
-		logger.Errorf("invalid resource key: %s", key)
-		return nil
-	}
-
-	// If our controller has configuration state, we'd "freeze" it and
-	// attach the frozen configuration to the context.
-	//    ctx = r.configStore.ToContext(ctx)
-
-	// Get the resource with this namespace/name.
-	original, err := r.Lister.AddressableServices(namespace).Get(name)
-	if apierrs.IsNotFound(err) {
-		// The resource may no longer exist, in which case we stop processing.
-		logger.Errorf("resource %q no longer exists", key)
-		return nil
-	} else if err != nil {
-		return err
-	}
-	// Don't modify the informers copy.
-	resource := original.DeepCopy()
-
-	// Reconcile this copy of the resource and then write back any status
-	// updates regardless of whether the reconciliation errored out.
-	reconcileErr := r.reconcile(ctx, resource)
-	if equality.Semantic.DeepEqual(original.Status, resource.Status) {
-		// If we didn't change anything then don't call updateStatus.
-		// This is important because the copy we loaded from the informer's
-		// cache may be stale and we don't want to overwrite a prior update
-		// to status with this stale state.
-	} else if _, err = r.updateStatus(resource); err != nil {
-		logger.Warnw("Failed to update resource status", zap.Error(err))
-		r.Recorder.Eventf(resource, corev1.EventTypeWarning, "UpdateFailed",
-			"Failed to update status for %q: %v", resource.Name, err)
-		return err
-	}
-	if reconcileErr != nil {
-		r.Recorder.Event(resource, corev1.EventTypeWarning, "InternalError", reconcileErr.Error())
-	}
-	return reconcileErr
-}
+// Check that our Reconciler implements reconciler.Interface
+var _ Interface = (*Reconciler)(nil)
 
-func (r *Reconciler) reconcile(ctx context.Context, asvc *v1alpha1.AddressableService) error {
+// ReconcileKind implements addressableservice Interface.ReconcileKind.
+func (r *Reconciler) ReconcileKind(ctx context.Context, asvc *v1alpha1.AddressableService) error {
 	if asvc.GetDeletionTimestamp() != nil {
```

Comment:
This is more boilerplate that we should elide.

Perhaps in addition to code-gen'ing `Interface` we can have a `Finalizer` interface with `FinalizeKind(ctx, asvc)`.

The signature to NewImpl would remain unchanged, but when the argument to `NewImpl`'s second argument implements this, it gets called?